Don't hide cursor on modifier press

Dav1dde opened the pull request:
Fixes: #2761.

chrisduerr commented on CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Preserve selection on non-LMB or mouse mode clicks
 - Wayland client side decorations are now based on config colorscheme
 - Low resolution window decoration icon on Windows
+- Don't hide cursor on modifier press (mouse.hide_when_typing setting)
```
```suggestion
- Don't hide cursor on modifier press with `mouse.hide_when_typing` enabled
```

This is about the same length but reads a bit nicer I think.

chrisduerr commented on alacritty/src/event.rs:
```diff
@@ -554,14 +554,23 @@ impl<N: Notify + OnResize> Processor<N> {
+                    WindowEvent::ReceivedCharacter(c) => {
+                        processor.received_char(c);
+                        // Hide cursor while typing
+                        if processor.ctx.config.ui_config.mouse.hide_when_typing
+                            && !*processor.ctx.suppress_chars()
+                        {
+                            processor.ctx.window.set_mouse_visible(false);
                         }
```
Why do we do this here instead of the `received_char` method? Seems like this would just duplicate the `suppress_chars` check, which seems inefficient.

chrisduerr commented on alacritty/src/event.rs:
```diff
@@ -554,14 +554,23 @@ impl<N: Notify + OnResize> Processor<N> {
-                                processor.ctx.window.set_mouse_visible(false);
-                            }
+                        // Hide cursor while typing, when triggering a keybind that writes to pty
+                        if input.state == ElementState::Pressed
+                            && processor.ctx.config.ui_config.mouse.hide_when_typing
+                            && *processor.ctx.received_count() > 0
```
This is not what `received_count` is for and it will break the existing `received_char` method.

Why is this not done directly in `input.rs`?

Dav1dde commented:
@chrisduerr thanks for the pointers, I moved the code over to `input.rs`. 

Maybe we should check for more actions than just `Action::Esc`, e.g. pasting does not hide the cursor (which feels okay for me).

I also split the initial check for suppress and vi mode in `received_char`, so the cursor is hidden if we receive a char in vi mode, I think that makes sense?

chrisduerr commented:
> Should we hide cursor on Vi mode actions, like moving around, etc.

We probably should.

Dav1dde commented:
> > Should we hide cursor on Vi mode actions, like moving around, etc.
> 
> We probably should.

The current implementation does that, or am I missing something?

kchibisov commented:
When I've checked it wasn't doing that for me, I can recheck though.

chrisduerr commented on alacritty/src/input.rs:
```diff
@@ -710,7 +710,15 @@ impl<'a, T: EventListener, A: ActionContext<T>> Processor<'a, T, A> {
     /// Process a received character.
     pub fn received_char(&mut self, c: char) {
-        if *self.ctx.suppress_chars() || self.ctx.terminal().mode().contains(TermMode::VI) {
+        if *self.ctx.suppress_chars() {
+            return;
+        }
+
+        if self.ctx.config().ui_config.mouse.hide_when_typing {
+            self.ctx.window_mut().set_mouse_visible(false);
+        }
+
+        if self.ctx.terminal().mode().contains(TermMode::VI) {
```
I'm not a huge fan of this approach, it's kinda messy. 

I think it would be better to explicitly hide the mouse courser when the `ViMotion` command is triggered.

Dav1dde replied:
This here exists for keys pressed that are *not* bound to an action. If an action is triggered `suppress_chars` is `true`. But if there is no action and we're in vi mode the cursor gets hidden, if that behaviour is not desired, I can "merge" the `suppress_chars` and vi mode check back together.

kchibisov replied:
I think we should, it'll also simplify Vi mode cursor hide handling.

Dav1dde replied:
Merged

chrisduerr commented on alacritty/src/input.rs:
```diff
@@ -768,6 +776,19 @@ impl<'a, T: EventListener, A: ActionContext<T>> Processor<'a, T, A> {
                 // Don't suppress when there has been a `ReceiveChar` action
                 *suppress_chars.get_or_insert(true) &= binding.action != Action::ReceiveChar;
+
+                // Hide the cursor if necessary
+                if self.ctx.config().ui_config.mouse.hide_when_typing {
+                    let hide_cursor = match binding.action {
+                        Action::Esc(_) => true,
+                        Action::ViMotion(_) => true,
+                        _ => false
+                    };
+
+                    if hide_cursor {
+                        self.ctx.window_mut().set_mouse_visible(false);
+                    }
+                }
```
Why does this need to be done here instead of the action handlers?

Dav1dde replied:
Good point, moved

Dav1dde commented:
Test failure seems unrelated:

```
The command "ci/script.sh" exited with 101.
error: failed to run custom build command for `winpty v0.2.0`
```
